### PR TITLE
Bucket Logging: BG worker to upload bucket logs

### PR DESCRIPTION
**Pull request by aspandey**
Bucket logs will be collected and placed on /log/noobaa_bucket_logs/ directory. The BG worker, BucketLogUploader,  will keep scanning this directory and upload the object on a log bucket.
This log bucket name will be fetched from the log object name.

 

### Testing Instructions:
Once the core pod is UP and running , exec into it and create  /log/noobaa_bucket_logs/ directory.

Create some files with bucket name which we have created using nb command.
if you have created bucket like this - 

nb api bucket_api create_bucket '{
  "name": "log.bucket"
 }'

Then create files as follows - 

echo "Bucket logging " >> /log/noobaa_bucket_logs/log.bucket-1.log

After some time - config.BUCKET_LOG_UPLOADER_DELAY = 5 * 60 * 100; - 
This log will be uploaded to log.bucket.

### CODE REVIEW

**Comment by aspandey** [comment]
@dannyzaken , @romayalon, @liranmauda 
This is a working PR. However, there are few things which we still need to be decided and make changes, for example what should be the delay of scanning , naming convention of log files and all.
I would like your initial comments on this patch.

File: src/server/bg_services/bucket_logs_upload.js [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+const _ = require('lodash');
+const system_store = require('../system_services/system_store').get_instance();
+const dbg = require('../../util/debug_module')(__filename);
+const system_utils = require('../utils/system_utils');
+const config = require('../../../config');
+const P = require('../../util/promise');
+const cloud_utils = require('../../util/cloud_utils');
+const fs = require('fs');
+
+const BUCKET_LOGS_PATH = '/log/noobaa_bucket_logs/';
```
Comment (liranmauda) [inline src/server/bg_services/bucket_logs_upload.js:13]: What will happen if this path does not exist?

File: src/server/bg_services/bucket_logs_upload.js [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+     */
+    constructor({ name, client }) {
+        this.name = name;
+        this.client = client;
+        this.noobaa_connection = undefined;
+    }
+
+    async run_batch() {
+        if (!this._can_run()) return;
+        dbg.log0('Bucket_log_uploader: starting scanning bucket logging objects');
```
Comment (liranmauda) [inline src/server/bg_services/bucket_logs_upload.js:31]:
```suggestion
        dbg.log0('Bucket_log_uploader: scanning bucket logging objects');
```

File: src/server/bg_services/bucket_logs_upload.js [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        const buckets = [];
+        if (!log_objects) return buckets;
+        for (const file of log_objects) {
+            const bucket_name = file.split('-')[0];
+            buckets.push({
+                bucket_name: bucket_name,
+                log_object_name: file,
+            });
+        }
+        return buckets;
```
Comment (liranmauda) [inline src/server/bg_services/bucket_logs_upload.js:67]:
```suggestion
        if (!log_objects) return [];
        const buckets = log_objects.map(file => {
            const bucket_name = file.split('-')[0];
            return {
                bucket_name,
                log_object_name: file,
            };
        });

        return buckets;
```

**Review by liranmauda — COMMENTED** [review]
Few comment:
1. When we use dbg, we don't need to specify the file name.
    for example: instead of writing  `dbg.log0('Bucket_log_uploader: system_store did not finish initial load'); ` we should write:  `dbg.log0('system_store did not finish initial load'); `
2. please add javadoc for typescript on every function.

File: config.js [hunk 4]
```diff
@@ -624,6 +624,13 @@ config.BUCKET_LOG_REPLICATOR_DELAY = 5 * 60 * 1000;
 config.AZURE_QUERY_TRUNCATION_MAX_SIZE_IN_BITS = 10 * 1024 * 1024;
 config.BUCKET_DIFF_FOR_REPLICATION = true;
 
+////////////////////////////////
+//      BUCKET LOGGING        //
+////////////////////////////////
+
+config.BUCKET_LOG_UPLOAD_ENABLED = true;
+config.BUCKET_LOG_UPLOADER_DELAY = 5 * 60 * 100;
```
[the remark this answers is not in the document]
Comment (aspandey) [reply config.js]: Done.

File: src/server/bg_services/bucket_logs_upload.js (hunk 1 above)
Comment (aspandey) [reply src/server/bg_services/bucket_logs_upload.js:13]: I think it should just throw an exception and continue..

File: src/server/bg_services/bucket_logs_upload.js (hunk 2 above)
Comment (aspandey) [reply src/server/bg_services/bucket_logs_upload.js]: done

File: src/server/bg_services/bucket_logs_upload.js (hunk 3 above)
Comment (aspandey) [reply src/server/bg_services/bucket_logs_upload.js]: done

File: src/server/system_services/schemas/bucket_schema.js [hunk 5]
```diff
@@ -277,8 +277,11 @@ module.exports = {
         },
         logging: {
             type: 'object',
-            required: ['log_bucket', 'log_prefix'],
+            required: ['name', 'log_bucket', 'log_prefix'],
             properties: {
+                name: {
+                    $ref: 'common_api#/definitions/bucket_name',
+                },
```
Comment (liranmauda) [inline src/server/system_services/schemas/bucket_schema.js:284]: Let's take it out into a different PR.

File: src/server/bg_services/bucket_logs_upload.js [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+
+    /**
+    * Read the directory BUCKET_LOGS_PATH
+    * and get the list of all log objects. After getting
+    * name of the bucket it uploades the log objects to the log bucket
+    * and creates list of buckets and respective log objects.
+    **/
+    async get_and_upload_bucket_log() {
+        let log_objects = null;
+        let log_file_path = null;
```
Comment (liranmauda) [inline src/server/bg_services/bucket_logs_upload.js:137]: decleration of variable does not require to be initiated with null

Comment (aspandey) [reply src/server/bg_services/bucket_logs_upload.js]: done

File: src/server/system_services/schemas/bucket_schema.js (hunk 5 above)
Comment (aspandey) [reply src/server/system_services/schemas/bucket_schema.js]: done

**Review by liranmauda — APPROVED** [review]
LGTM

**Pull request merged**